test/e2e/apps/statefulset.go

```diff
@@ -1400,6 +1465,175 @@ var _ = SIGDescribe("StatefulSet", func() {
+			err = wait.PollUntilContextTimeout(ctx, e2estatefulset.StatefulSetPoll, e2estatefulset.StatefulSetTimeout, true, func(ctx context.Context) (bool, error) {
+				for _, i := range []int{0, 2} {
+					claim, err := c.CoreV1().PersistentVolumeClaims(ns).Get(ctx, claimNames[i], metav1.GetOptions{})
+					if err != nil {
+						return false, nil // retry
+					}
+					if !reflect.DeepEqual(claim.GetOwnerReferences(), expectedOwnerRef) {
+						return false, nil // retry
+					}
+				}
+				claim, err := c.CoreV1().PersistentVolumeClaims(ns).Get(ctx, claimNames[3], metav1.GetOptions{})
+				if err != nil {
+					return false, nil // retry
+				}
+				if !reflect.DeepEqual(claim.GetOwnerReferences(), expectedStaleRef) {
+					return false, nil // retry
+				}
+				return true, nil // found them all!
+			})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Remove controller flag from claim 0")
+			claim, err = c.CoreV1().PersistentVolumeClaims(ns).Get(ctx, claimNames[0], metav1.GetOptions{})
+			framework.ExpectNoError(err)
+			claim.SetOwnerReferences([]metav1.OwnerReference{
+				{
+					APIVersion:         "apps/v1",
+					Kind:               "StatefulSet",
+					Name:               ssName,
+					UID:                ss.GetUID(),
+					Controller:         ptr.To(false),
+					BlockOwnerDeletion: ptr.To(true),
+				},
+			})
+			_, err = c.CoreV1().PersistentVolumeClaims(ns).Update(ctx, claim, metav1.UpdateOptions{})
+			framework.ExpectNoError(err)
+			ginkgo.By("Update statefulset to provoke a reconcile")
+			ss, err = c.AppsV1().StatefulSets(ns).Get(ctx, ssName, metav1.GetOptions{})
+			framework.ExpectNoError(err)
+			ss.Spec.PersistentVolumeClaimRetentionPolicy = &appsv1.StatefulSetPersistentVolumeClaimRetentionPolicy{
+				WhenDeleted: appsv1.DeletePersistentVolumeClaimRetentionPolicyType,
+				WhenScaled:  appsv1.DeletePersistentVolumeClaimRetentionPolicyType,
+			}
+			ss, err = c.AppsV1().StatefulSets(ns).Update(ctx, ss, metav1.UpdateOptions{})
+			framework.ExpectNoError(err)
+			ginkgo.By("Expect controller flag for claim 0 to reconcile back to true")
+			err = wait.PollUntilContextTimeout(ctx, e2estatefulset.StatefulSetPoll, e2estatefulset.StatefulSetTimeout, true, func(ctx context.Context) (bool, error) {
+				claim, err := c.CoreV1().PersistentVolumeClaims(ns).Get(ctx, claimNames[0], metav1.GetOptions{})
+				if err != nil {
+					return false, nil // retry
+				}
+				if reflect.DeepEqual(claim.GetOwnerReferences(), expectedOwnerRef) {
+					return true, nil // success!
+				}
+				return false, nil // retry
+			})
+			framework.ExpectNoError(err)
+
+			// Note 3 has a finalizer still.
+			ginkgo.By("Delete the stateful set and wait for claims 0 and 2 but not 1 and 3 to disappear")
+			err = c.AppsV1().StatefulSets(ns).Delete(ctx, ssName, metav1.DeleteOptions{})
+			framework.ExpectNoError(err)
+			err = verifyStatefulSetPVCsExist(ctx, c, ss, []int{1, 3})
```

Comment:
how come 1 is still there if the expectedExternalRef doesn't have the controller flag? Am I missing something?

Reply:
An external owner ref is a stale owner ref, see stateful_set_utils.go:isClainOwnerUpToDate.

Reply:
I understand it better now. It reconciles normally and still has both refs: `slices.Concat(expectedExternalRef, expectedOwnerRef)`

It does not get deleted because of the first one. Unfortunately because we create a non-existent reference we fire the following errors in the KCM garbage collector:

```
E0501 00:27:21.573617  154898 garbagecollector.go:397] "error syncing item" err="unable to get REST mapping for unknown/v1/randomOwner." logger="garbage-collector-controller" item="[v1/PersistentVolumeClaim, namespace: statefulset-7724, name: datadir-ss-1, uid: 6f062806-1064-4df6-a6da-63c2cd3d7929]"
```

which I think is rather unfortunate, and it would be better if we could prevent that by creating an existing reference or at least an existing kind.

Reply:
+1. Can we also properly comment out why the PVCs are not getting deleted at the end?

Reply:
done